sync2aptly/src/lib.rs

```diff
@@ -230,12 +231,51 @@ impl Display for OriginLocation {
     }
 }
 
+// Unfortunately, we can't easily rely on once_cell's Lazy here, because it
+// would end up with an &Result<...>, which can't be trivially handled as a
+// non-ref Result<...> without cloning...which we can't do, because ErrReport
+// isn't Clone.
+type LazyVersionFn = Box<dyn Fn() -> Result<PackageVersion> + Send + Sync>;
+
+pub struct LazyVersion {
+    version: OnceCell<PackageVersion>,
+    compute: Option<LazyVersionFn>,
+}
+
+impl LazyVersion {
+    pub fn new(compute: LazyVersionFn) -> Self {
+        Self {
+            version: OnceCell::new(),
+            compute: Some(compute),
+        }
+    }
+
+    pub fn with_value(version: PackageVersion) -> Self {
+        Self {
+            version: OnceCell::with_value(version),
+            compute: None,
+        }
+    }
+
+    pub fn get(&self) -> Result<&PackageVersion> {
+        self.version
+            .get_or_try_init(|| self.compute.as_ref().unwrap()())
+    }
+}
+
+impl std::fmt::Debug for LazyVersion {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.version.fmt(f)
+    }
+}
+
 #[derive(Debug)]
 pub struct OriginDeb {
     pub package: PackageName,
+    pub version: LazyVersion,
```

Comment:
This is not a merge blocker for me;  but why is this dynamic for each consumer as we'll likely only have two cases:
* The version is fully determined initially (e.g. when parsing from a packaging file)
* The version could be made more specificly (e.g. epoch) which is likely always going to be via parsing the .deb

In other words sync2aptly could just hardcode parsing a local .deb as you currently do in obs2aptly and probably for now have an unimplemented for the remote case. 

I don't really see the value in requiring each sync2aptly to provide this implementation detail

Reply:
Using once_cell seemed advantageous because there are some code paths where we read the version multiple times, and making the compute function caller-provided wasn't much of a stretch on top; that being said, I'm not particularly committed to this.